python/cudf/cudf/core/frame.py

```diff
@@ -235,9 +241,67 @@ def _concat(cls, objs, axis=0, ignore_index=False, sort=False):
                     result_index_length += len(obj)
                     empty_has_index = empty_has_index or len(obj) > 0
 
-        # Get a list of the unique table column names
-        names = [name for f in objs for name in f._column_names]
-        names = OrderedDict.fromkeys(names).keys()
+        if join == "inner":
+            old_names = [name for obj in objs for name in obj._column_names]
+            setobjs = [set(obj._column_names) for obj in objs]
+            i = set(setobjs[0])
+            for x in setobjs[1:]:
+                i = i & set(x)
+            names_not_in_all = [name for name in old_names if name not in i]
+            names = [name for name in old_names if name in i]
+            names = OrderedDict.fromkeys(names).keys()
+            old_dtypes = [dt for obj in objs for dt in obj.dtypes]
+
+            if axis == 0:
+                if num_empty_input_frames > 0 or old_names == names_not_in_all:
```

Comment:
```suggestion
                if num_empty_input_frames > 0 or len(old_names) == len(names_not_in_all):
```

I think this should be length check instead of element wise comparison